### ISSUE TITLE
Unable to delete authorized RP application

### ISSUE THREAD

**Issue by truongnmt** [doorkeeper-gem/doorkeeper-openid_connect#82]
### Steps to reproduce
- Create an application
- User authorizes application
- Go to doorkeeper admin page and delete application

### Expected behavior
Able to delete application

### Actual behavior
Unable to delete application:
> Mysql2::Error - Cannot delete or update a parent row: a foreign key constraint fails (sample_app.oauth_openid_requests, CONSTRAINT fk_rails_77114b3b09 FOREIGN KEY (access_grant_id) REFERENCES oauth_access_grants (id)):

### Reason
There are two problems.
First the problems above about Mysql, on `doorkeeper-openid-connect` table migration, I found that foreign key is added:
https://github.com/doorkeeper-gem/doorkeeper-openid_connect/blob/302575597532ecaf5a4ae8000b2c387f7d08371a/lib/generators/doorkeeper/openid_connect/templates/migration.rb.erb#L8-L12

It's off-topic but I'm not understand the purpose to use foreign key, as all the validation can be done on Rails. 

There are associations between: 
- `Doorkeeper::Application` 
- `Doorkeeper::OpenidConnect::AccessGrant`  
- `Doorkeeper::OpenidConnect::Request`

Delete `Application` will delete `AccessGrant` and delete `Request`. However due to foreign key, `AccessGrant` can't be delete without delete associated `Request` first.

That's one problems, we can solve that by deleting foreign key in `templates/migration.rb.erb`, for existing application we can use migration:
```ruby
def up
  remove_foreign_key :oauth_openid_requests, :oauth_access_grants
end

def down
  add_foreign_key(
      :oauth_openid_requests,
      :oauth_access_grants,
      column: :access_grant_id
  )
end
```

No error occurs. However, another problem is that associated `openid_request` are not deleted. To get to the reason, first let's see the model association:

Doorkeeper::Application
https://github.com/doorkeeper-gem/doorkeeper/blob/45b6ce568523a5e2f2cdc9c3ed60254d7891f8dc/lib/doorkeeper/orm/active_record/application.rb#L3-L11
```ruby
class Doorkeeper::Application
    has_many :access_grants, dependent: :delete_all, class_name: "Doorkeeper::AccessGrant"
    has_many :access_tokens, dependent: :delete_all, class_name: "Doorkeeper::AccessToken"
end
```

Doorkeeper::OpenidConnect::AccessGrant
https://github.com/doorkeeper-gem/doorkeeper-openid_connect/blob/302575597532ecaf5a4ae8000b2c387f7d08371a/lib/doorkeeper/openid_connect/orm/active_record/access_grant.rb#L1-L10

To recap:
```
Application has_many :access_grants, dependent: :delete_all
AccessGrant has_one :openid_request, dependent: :delete
```

When application is deleted, access_grants also are deleted, then openid_request should be deleted as well. However, because we are using `dependent: :delete_all` in `Doorkeeper::Application` so callback stop after delete `:access_grants`. If I change to `dependent: :destroy`, both `:access_grants` **and** `:openid_request` are deleted.

**Comment by truongnmt** [comment]
Sorry for late reply! Seem like @syakovyn already opened one. 
Btw I'm happy to work on such PR.

**Comment by toupeira** [comment]
As discussed in https://github.com/doorkeeper-gem/doorkeeper/issues/1311, it's tricky to switch to `dependent: destroy_all` in `Doorkeeper::Application` due to performance concerns.

But I'm happy to accept a PR which implements a monkey-patch along the lines of https://github.com/doorkeeper-gem/doorkeeper/issues/1311#issuecomment-537909914.

**Comment by toupeira** [comment]
This has come up on GitLab as well: https://gitlab.com/gitlab-org/gitlab/merge_requests/19617

The solution there was to use `ON DELETE CASCADE`, which is much cleaner and seems to be supported on PostgreSQL/MySQL/SQLite. So I'm planning to add a migration to this gem soon to change the foreign key.
